src/charm.py

```diff
@@ -119,6 +122,36 @@ def on_config_changed_handler(event, fw_adapter, state):
         pod_is_ready = isinstance(juju_unit_status, ActiveStatus)
         time.sleep(1)
 
+    if not state.is_started:
+        state.is_started = True
+    else:
+        logging.debug("Call HTTP config reload")
+        try:
+            fw_adapter.set_unit_status(MaintenanceStatus(
+                "Waiting for config parameters propagation"
+            ))
+            expected_config = build_prometheus_config(fw_adapter.get_config())
+            logging.debug(
+                "Awaiting Prom config to become: {0}".format(expected_config)
+            )
+
+            while True:
+                # Wait, until ConfigMap changes will be propagated
+                prometheus_config_reload(juju_model, juju_app)
+                new_config_applied = prometheus_ensure_config_propagated(
+                    juju_model, juju_app, expected_config
+                )
+                if new_config_applied:
+                    break
+                time.sleep(5)
+
+            logging.debug("Config reloaded")
+            fw_adapter.set_unit_status(ActiveStatus())
+        except CharmError as e:
+            error_msg = "Config reload error: {0}".format(e)
+            logging.error(error_msg)
+            fw_adapter.set_unit_status(BlockedStatus(error_msg))
+
```

Comment:
Much of this logic needs to be moved to one or more stateless functions in `domain.py`. The logic in `charm.py` should always be limited to just dispatching other functions otherwise it will start to become bloated. Another side effect of putting business logic here is that you will have a hard time writing your unit tests. Either that or the unit tests will be hard to read and maintain. Please see [Charmed LMA Operators Architecture](https://docs.google.com/document/d/1V5cA9D1YN8WGEClpLhUwQt2dYrg5VZEY99LZiE6Mx_A/edit?usp=sharing) for guidance. Don't hesitate to ask questions here or in that document if anything is unclear.

Reply:
Alright - moved to `domain.py` :-)